File: practice/msa/board-service/src/main/java/com/example/boardservice/controller/BoardApiController.java
package com.example.boardservice.controller;

import com.example.boardservice.domain.entity.Board;
import com.example.boardservice.dto.*;
import com.example.boardservice.mapper.BoardMapper;
import com.example.boardservice.service.BoardService;
import com.example.boardservice.service.FileService;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/boards")
public class BoardApiController {
    private final BoardService boardService;
    private final FileService fileService;
    private final BoardMapper boardMapper;

    @GetMapping("/search")
    public Page<BoardListItemResponseDto> searchBoards(
            @ModelAttribute BoardSearchRequestDto dto,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        return boardService.searchBoards(dto, PageRequest.of(page - 1, size));
    }

    @GetMapping("/{id}/with-comments")
    public BoardWithCommentsResponseDto getBoardWithComments(@PathVariable("id") long id) {
        Board board = boardService.getBoardWithComments(id);
        return boardMapper.toBoardWithCommentsResponseDto(board);
    }

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public void saveBoard(@ModelAttribute BoardWriteRequestDto dto) {
        boardService.saveBoard(dto.getUserId(), dto.getTitle(), dto.getContent(), dto.getFile());
    }

    @GetMapping("/{id}")
    public BoardDetailResponseDto getBoardDetail(@PathVariable long id) {
        Board boardDetail = boardService.getBoardDetail(id);

        return BoardDetailResponseDto.builder()
                .title(boardDetail.getTitle())
                .content(boardDetail.getContent())
                .filePath(boardDetail.getFilePath())
                .created(boardDetail.getCreated())
                .userId(boardDetail.getUserId())
                .build();
    }

    @PutMapping(value = "/{id}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public void updateBoard(
            @PathVariable long id,
            @ModelAttribute BoardUpdateRequestDto dto
    ) {
        boardService.updateBoard(id, dto);
    }

    @DeleteMapping("/{id}")
    public void deleteBoard(
            @PathVariable long id,
            @RequestBody BoardDeleteRequestDto dto
    ) {
        boardService.deleteBoard(id, dto);
    }

    @GetMapping("/file/download/{fileName}")
    public ResponseEntity<Resource> downloadFile(@PathVariable("fileName") String fileName) {
        Resource resource = fileService.downloadFile(fileName);

        String encodedFileName = URLEncoder.encode(resource.getFilename(), StandardCharsets.UTF_8)
                .replaceAll("\\+", "%20");

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_OCTET_STREAM)
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename*=UTF-8''" + encodedFileName)
                .body(resource);
    }

    @GetMapping("/stats/authors")
    public List<BoardAuthorStatsResponseDto> getAuthors(
            @RequestParam long minCount
    ) {
        return boardService.getAuthorStats(minCount);
    }
}

File: practice/springboot/board/kotlin-board/src/main/kotlin/com/example/kotlinboard/controller/BoardApiController.kt
package com.example.kotlinboard.controller

import com.example.kotlinboard.dto.BoardCreateRequest
import com.example.kotlinboard.dto.BoardPageResponse
import com.example.kotlinboard.dto.BoardResponse
import com.example.kotlinboard.dto.BoardUpdateRequest
import com.example.kotlinboard.service.BoardService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/boards")
class BoardApiController(private val boardService: BoardService) {
    @GetMapping
    fun getBoards(
        @RequestParam(value = "page", defaultValue = "1") page: Int,
        @RequestParam(value = "size", defaultValue = "10") size: Int,
        @RequestParam(value = "keyword", required = false) keyword: String?
    ): BoardPageResponse = BoardPageResponse.from(boardService.getBoards(page, size, keyword))

    @PostMapping
    fun createBoard(@RequestBody request: BoardCreateRequest) {
        boardService.createBoard(request)
    }

    @GetMapping("/{id}")
    fun getBoard(@PathVariable("id") id: Long): ResponseEntity<BoardResponse> {
        val board = boardService.getBoard(id) ?: return ResponseEntity.notFound().build()

        return ResponseEntity.ok(BoardResponse.from(board))
    }

    @PutMapping("/{id}")
    fun updateBoard(
        @PathVariable("id") id: Long,
        @RequestBody request: BoardUpdateRequest
    ) {
        boardService.updateBoard(id, request)
    }

    @DeleteMapping("/{id}")
    fun deleteBoard(@PathVariable("id") id: Long) {
        boardService.deleteBoard(id)
    }
}